File: frontend/src/lib/schemas/page-config.ts
import { z } from "zod";
import { HeaderContentSchema } from "./content/header";
import { HeroContentSchema } from "./content/hero";
import { LogosContentSchema } from "./content/logos";
import { FeaturesContentSchema } from "./content/features";
import { SocialProofContentSchema } from "./content/social-proof";
import { StatsContentSchema } from "./content/stats";
import { PricingContentSchema } from "./content/pricing";
import { FaqContentSchema } from "./content/faq";
import { CtaContentSchema } from "./content/cta";
import { FooterContentSchema } from "./content/footer";

// --- Section types ---

export const SECTION_TYPES = [
  "header",
  "hero",
  "logos",
  "features",
  "social-proof",
  "stats",
  "pricing",
  "faq",
  "cta",
  "footer",
] as const;

export type SectionType = (typeof SECTION_TYPES)[number];

// Map section types to their content schemas
export const contentSchemaMap: Record<SectionType, z.ZodType> = {
  header: HeaderContentSchema,
  hero: HeroContentSchema,
  logos: LogosContentSchema,
  features: FeaturesContentSchema,
  "social-proof": SocialProofContentSchema,
  stats: StatsContentSchema,
  pricing: PricingContentSchema,
  faq: FaqContentSchema,
  cta: CtaContentSchema,
  footer: FooterContentSchema,
};

// --- Theme ---

export const ThemeSchema = z.object({
  primary: z.string().default("221.2 83.2% 53.3%"),
  primaryForeground: z.string().default("210 40% 98%"),
  secondary: z.string().default("210 40% 96.1%"),
  secondaryForeground: z.string().default("222.2 47.4% 11.2%"),
  background: z.string().default("0 0% 100%"),
  foreground: z.string().default("240 10% 3.9%"),
  muted: z.string().default("210 40% 96.1%"),
  mutedForeground: z.string().default("215.4 16.3% 46.9%"),
  accent: z.string().default("210 40% 96.1%"),
  accentForeground: z.string().default("222.2 47.4% 11.2%"),
  border: z.string().default("214.3 31.8% 91.4%"),
  ring: z.string().default("221.2 83.2% 53.3%"),
  destructive: z.string().default("0 84.2% 60.2%"),
  destructiveForeground: z.string().default("210 40% 98%"),
  radius: z.string().default("0.5rem"),
  fontHeading: z.string().default("Inter"),
  fontBody: z.string().default("Inter"),
});

export type Theme = z.infer<typeof ThemeSchema>;

// --- Dark theme overrides (used for .dark class) ---

export const DarkThemeOverrides: Partial<Theme> = {
  background: "222.2 84% 4.9%",
  foreground: "210 40% 98%",
  secondary: "217.2 32.6% 17.5%",
  secondaryForeground: "210 40% 98%",
  muted: "217.2 32.6% 17.5%",
  mutedForeground: "215 20.2% 65.1%",
  accent: "217.2 32.6% 17.5%",
  accentForeground: "210 40% 98%",
  border: "217.2 32.6% 17.5%",
};

// --- Section ---

export const ModeSchema = z.enum(["light", "dark"]);
export type Mode = z.infer<typeof ModeSchema>;

export const SectionSchema = z.object({
  id: z.string().uuid(),
  type: z.enum(SECTION_TYPES),
  variant: z.string(),
  mode: ModeSchema.default("light"),
  content: z.record(z.string(), z.unknown()),
});

export type Section = z.infer<typeof SectionSchema>;

// --- Page Metadata ---

export const PageMetadataSchema = z.object({
  title: z.string().default(""),
  description: z.string().default(""),
  ogImage: z.string().optional(),
  favicon: z.string().optional(),
});

export type PageMetadata = z.infer<typeof PageMetadataSchema>;

// --- Page Config (top-level) ---

export const PageConfigSchema = z.object({
  id: z.string().uuid(),
  name: z.string(),
  theme: ThemeSchema.default(() => ThemeSchema.parse({})),
  sections: z.array(SectionSchema).default([]),
  metadata: PageMetadataSchema.default(() => PageMetadataSchema.parse({})),
  version: z.number().int().default(1),
});

export type PageConfig = z.infer<typeof PageConfigSchema>;

File: backend/app/schemas/page_config.py
"""Pydantic mirrors of the frontend Zod schemas. Frontend Zod is source of truth."""

from __future__ import annotations

from pydantic import BaseModel, Field

# --- Content schemas (one per section type) ---


class NavItem(BaseModel):
    label: str
    href: str


class LogoRef(BaseModel):
    text: str
    href: str = "/"


class CtaButton(BaseModel):
    label: str
    href: str


class ImageRef(BaseModel):
    src: str
    alt: str


class HeaderContent(BaseModel):
    logo: LogoRef
    navItems: list[NavItem] = []
    cta: CtaButton | None = None


class HeroContent(BaseModel):
    headline: str
    subheadline: str | None = None
    primaryCta: CtaButton | None = None
    secondaryCta: CtaButton | None = None
    image: ImageRef | None = None
    badge: str | None = None


class LogoItem(BaseModel):
    name: str
    src: str
    href: str | None = None


class LogosContent(BaseModel):
    title: str | None = None
    logos: list[LogoItem] = []


class FeatureItem(BaseModel):
    icon: str | None = None
    title: str
    description: str
    image: ImageRef | None = None


class FeaturesContent(BaseModel):
    tagline: str | None = None
    headline: str
    description: str | None = None
    features: list[FeatureItem] = Field(min_length=1)


class Testimonial(BaseModel):
    quote: str
    author: str
    role: str | None = None
    company: str | None = None
    avatar: str | None = None
    rating: int | None = Field(None, ge=1, le=5)


class SocialProofContent(BaseModel):
    tagline: str | None = None
    headline: str | None = None
    testimonials: list[Testimonial] = Field(min_length=1)


class StatItem(BaseModel):
    value: str
    label: str
    description: str | None = None


class StatsContent(BaseModel):
    tagline: str | None = None
    headline: str | None = None
    stats: list[StatItem] = Field(min_length=1)


class PricingFeature(BaseModel):
    text: str
    included: bool = True


class PricingPlan(BaseModel):
    name: str
    description: str | None = None
    price: str
    period: str = "/month"
    features: list[PricingFeature] = []
    cta: CtaButton
    highlighted: bool = False
    badge: str | None = None


class PricingContent(BaseModel):
    tagline: str | None = None
    headline: str
    description: str | None = None
    plans: list[PricingPlan] = Field(min_length=1)


class FaqItem(BaseModel):
    question: str
    answer: str
    category: str | None = None


class FaqContent(BaseModel):
    tagline: str | None = None
    headline: str
    description: str | None = None
    items: list[FaqItem] = Field(min_length=1)


class CtaContent(BaseModel):
    headline: str
    description: str | None = None
    primaryCta: CtaButton
    secondaryCta: CtaButton | None = None
    image: ImageRef | None = None


class FooterLink(BaseModel):
    label: str
    href: str


class FooterColumn(BaseModel):
    title: str
    links: list[FooterLink] = []


class SocialLink(BaseModel):
    platform: str
    href: str
    icon: str | None = None


class Newsletter(BaseModel):
    title: str
    description: str | None = None
    placeholder: str = "Enter your email"
    buttonLabel: str = "Subscribe"


class FooterContent(BaseModel):
    logo: LogoRef | None = None
    columns: list[FooterColumn] = []
    socialLinks: list[SocialLink] = []
    copyright: str | None = None
    newsletter: Newsletter | None = None


# --- Map type names to content models ---

CONTENT_MODEL_MAP: dict[str, type[BaseModel]] = {
    "header": HeaderContent,
    "hero": HeroContent,
    "logos": LogosContent,
    "features": FeaturesContent,
    "social-proof": SocialProofContent,
    "stats": StatsContent,
    "pricing": PricingContent,
    "faq": FaqContent,
    "cta": CtaContent,
    "footer": FooterContent,
}


# --- Theme ---

class ThemeConfig(BaseModel):
    primary: str = "221.2 83.2% 53.3%"
    primaryForeground: str = "210 40% 98%"
    secondary: str = "210 40% 96.1%"
    secondaryForeground: str = "222.2 47.4% 11.2%"
    background: str = "0 0% 100%"
    foreground: str = "240 10% 3.9%"
    muted: str = "210 40% 96.1%"
    mutedForeground: str = "215.4 16.3% 46.9%"
    accent: str = "210 40% 96.1%"
    accentForeground: str = "222.2 47.4% 11.2%"
    border: str = "214.3 31.8% 91.4%"
    ring: str = "221.2 83.2% 53.3%"
    destructive: str = "0 84.2% 60.2%"
    destructiveForeground: str = "210 40% 98%"
    radius: str = "0.5rem"
    fontHeading: str = "Inter"
    fontBody: str = "Inter"


# --- Section ---

SECTION_TYPES = [
    "header", "hero", "logos", "features", "social-proof",
    "stats", "pricing", "faq", "cta", "footer",
]


class SectionConfig(BaseModel):
    id: str
    type: str  # validated against SECTION_TYPES
    variant: str
    mode: str = "light"  # "light" | "dark"
    content: dict  # validated per-type using CONTENT_MODEL_MAP


# --- Page metadata ---

class PageMetadata(BaseModel):
    title: str = ""
    description: str = ""
    ogImage: str | None = None
    favicon: str | None = None


# --- Page config (top-level) ---

class PageConfig(BaseModel):
    id: str
    name: str
    theme: ThemeConfig = ThemeConfig()
    sections: list[SectionConfig] = []
    metadata: PageMetadata = PageMetadata()
    version: int = 1


def validate_page_config(config: dict) -> PageConfig:
    """Validate a raw dict as a PageConfig, including per-section content validation."""
    pc = PageConfig.model_validate(config)
    for section in pc.sections:
        if section.type in CONTENT_MODEL_MAP:
            model = CONTENT_MODEL_MAP[section.type]
            model.model_validate(section.content)
    return pc
